include/zephyr/bluetooth/buf.h

```diff
@@ -85,13 +85,39 @@ struct bt_buf_data {
 #define BT_BUF_ISO_RX_COUNT 0
 #endif /* CONFIG_BT_ISO */
 
+/* see Core Spec v6.0 vol.4 part E 7.4.5 */
+#define BT_BUF_ACL_RX_COUNT_MAX 65535
+#define BT_BUF_ACL_RX_COUNT_DEFAULT 6
+
+#if defined(CONFIG_BT_CONN) && defined(CONFIG_BT_HCI_HOST)
+ /* The host needs more ACL buffers than maximum ACL links. This is because of
+  * the way we re-assemble ACL packets into L2CAP PDUs.
+  *
+  * We keep around the first buffer (that comes from the driver) to do
+  * re-assembly into, and if all links are re-assembling, there will be no buffer
+  * available for the HCI driver to allocate from.
+  *
+  * Fixing it properly involves a re-design of the HCI driver interface.
+  *
+  * TODO: When CONFIG_BT_BUF_ACL_RX_COUNT is removed,
+  *       remove the MAX and only keep (CONFIG_BT_MAX_CONN + 1)
+  */
+#define BT_BUF_ACL_RX_COUNT (MAX(CONFIG_BT_BUF_ACL_RX_COUNT, (CONFIG_BT_MAX_CONN + 1)) + \
+			     CONFIG_BT_BUF_ACL_RX_COUNT_EXTRA)
```

Comment:
Maybe but I am not entirely sure because I did not remove the default (6). That mean that every application will get the warning even if they don't set the config themselves. It may be confusing for users, what do you think?

Reply:
Oh, I thought you modified the default to be `0`. Wouldn't that make more sense? That way apps which set an explicit value still work (but get a warning) but everyone else will just get the new automatically calculated value.

Reply:
I thought of setting it to zero but it may break apps that may use the config value even tho they don't set it explicitly. I don't know at we point we can say it's too many "may" 😅

Reply:
> I thought of setting it to zero but it may break apps that may use the config value even tho they don't set it explicitly. I don't know at we point we can say it's too many "may" 😅

My gut feeling is that this is something we can with a good conscience cover with a proper mention in `migration-guide-4.1.rst`.

Reply:
We discussed it with @alwa-nordic and agree with you. I added the warning, set the default to 0 and updated the migration guide. Tell me if you find the migration guide good enough (I also updated the commit message btw)